### ISSUE TITLE
Firmware Update fails on Nitrokey 3C

### ISSUE THREAD

**Issue by maaattes** [Nitrokey/pynitrokey#192]
My system:
Fedora 35
Python 3.10.2
pynitrokey installed via `pip install --user pynitrokey`

```console
[mattes@t480s ~]$ nitropy version
Nitrokey tool for Nitrokey FIDO2, Nitrokey Start, Nitrokey 3 & NetHSM
0.4.16
[mattes@t480s ~]$ nitropy nk3 fetch-update
Nitrokey tool for Nitrokey FIDO2, Nitrokey Start, Nitrokey 3 & NetHSM
Download v1.0.1: 100%|███████████████████████████| 304k/304k [00:00<00:00, 715kB/s]
Successfully downloaded firmware release v1.0.1 to ./firmware-nk3xn-lpc55-v1.0.1.sb2
[mattes@t480s ~]$ nitropy nk3 update firmware-nk3xn-lpc55-v1.0.1.sb2 
Nitrokey tool for Nitrokey FIDO2, Nitrokey Start, Nitrokey 3 & NetHSM
Current firmware version:  v1.0.0
Updated firmware version:  v1.0.1

Please do not remove the Nitrokey 3 or insert any other Nitrokey 3 devices during the update. Doing so may damage the Nitrokey 3.
Do you want to perform the firmware update now? [y/N]: y

Please press the touch button to reboot the device into bootloader mode ...

Traceback (most recent call last):
  File "/home/mattes/.local/bin/nitropy", line 8, in <module>
    sys.exit(nitropy())
  File "/home/mattes/.local/lib/python3.10/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/mattes/.local/lib/python3.10/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/mattes/.local/lib/python3.10/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/mattes/.local/lib/python3.10/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/mattes/.local/lib/python3.10/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/mattes/.local/lib/python3.10/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/mattes/.local/lib/python3.10/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/home/mattes/.local/lib/python3.10/site-packages/pynitrokey/cli/nk3/__init__.py", line 377, in update
    raise CliException(*msgs, exc)
  File "/home/mattes/.local/lib/python3.10/site-packages/pynitrokey/cli/exceptions.py", line 19, in __init__
    super().__init__("\n".join(messages))
TypeError: sequence item 2: expected str instance, McuBootConnectionError found
```

**Comment by maaattes** [comment]
> Hi! Thank you for the report. This error message should not be that unpleasant. As for the cause, please check if you have Udev rules installed:
> 
>     * https://docs.nitrokey.com/nitrokey3/linux/troubleshooting.html#nitrokey-is-not-detected-on-linux
I've copied the udev rules beforehand.

**Comment by robin-nitrokey** [comment]
@maaattes Can you please follow the steps described [in this comment](https://github.com/Nitrokey/pynitrokey/issues/167#issuecomment-1024921046) to make sure that the udev rules are applied correctly?

**Comment by maaattes** [comment]
Sure. Here are my results:
```console
[mattes@t480s ~]$ nitropy nk3 reboot --bootloader
[mattes@t480s ~]$ dmesg
…
[79865.086577] hid-generic 0003:20A0:42DD.000F: hiddev100,hidraw7: USB HID v1.00 Device [NXP SEMICONDUCTOR INC. USB COMPOSITE DEVICE] on usb-0000:3c:00.0-1/input0
[mattes@t480s ~]$ udevadm info --query=path /dev/hidraw7
/devices/pci0000:00/0000:00:1c.4/0000:04:00.0/0000:05:02.0/0000:3c:00.0/usb3/3-1/3-1:1.0/0003:20A0:42DD.000F/hidraw/hidraw7
[mattes@t480s ~]$ udevadm test /devices/pci0000:00/0000:00:1c.4/0000:04:00.0/0000:05:02.0/0000:3c:00.0/usb3/3-1/3-1:1.0/0003:20A0:42DD.000F/hidraw/hidraw7
This program is for debugging only, it does not run any program
specified by a RUN key. It may show incorrect results, because
some values may be different, or not available at a simulation run.

Trying to open "/etc/systemd/hwdb/hwdb.bin"...
Trying to open "/etc/udev/hwdb.bin"...
…

Reading rules file: /etc/udev/rules.d/41-nitrokey.rules
…
hidraw7: /usr/lib/udev/rules.d/60-fido-id.rules:5 Importing properties from results of 'fido_id'
hidraw7: Starting 'fido_id'
Successfully forked off '(spawn)' as PID 145516.
Skipping PR_SET_MM, as we don't have privileges.
hidraw7: 'fido_id'(err) 'Failed to get current device from environment: Invalid argument'
hidraw7: Process 'fido_id' failed with exit code 1.
hidraw7: /usr/lib/udev/rules.d/60-fido-id.rules:5 Command "fido_id" returned 1 (error), ignoring
hidraw7: Handling device node '/dev/hidraw7', devnum=c239:7
hidraw7: Preserve permissions of /dev/hidraw7, uid=0, gid=0, mode=0600
hidraw7: Failed to adjust timestamp of node /dev/hidraw7: Permission denied
DEVPATH=/devices/pci0000:00/0000:00:1c.4/0000:04:00.0/0000:05:02.0/0000:3c:00.0/usb3/3-1/3-1:1.0/0003:20A0:42DD.000F/hidraw/hidraw7
DEVNAME=/dev/hidraw7
MAJOR=239
MINOR=7
ACTION=add
SUBSYSTEM=hidraw
Unload module index
Unloaded link configuration context.
```

**Comment by robin-nitrokey** [comment]
Thank you.  Apparently, the udev rules are loaded but not applied.  Can you share the content of `/etc/udev/rules.d/41-nitrokey.rules` please?  Are you using systemd on your system?

**Comment by maaattes** [comment]
Yes, I'm using systemd.
I've just updated the udev rules so I can't give you the content of the "old" rules. But I'm pretty sure that these were up-to-date.
Running `nitropy nk3 test` gives me on error.

```console
Nitrokey tool for Nitrokey FIDO2, Nitrokey Start, Nitrokey 3 & NetHSM
Found 1 Nitrokey 3 device(s):
- Nitrokey 3 at /dev/hidraw7

Running tests for Nitrokey 3 at /dev/hidraw7

[1/3]	UUID query              	SUCCESS  	[not supported]
[2/3]	Firmware version query  	SUCCESS  	v1.0.0
Please press the touch button on the device ...
[3/3]	FIDO2                   	FAILURE  	PIN activated -- please set the --pin option

3 tests, 2 successful, 0 skipped, 1 failed

Summary: 1 device(s) tested, 0 successful, 1 failed

Critical error:
Test failed for 1 device(s)
```
Could this be the cause?

**Comment by robin-nitrokey** [comment]
No, the `nitropy nk3 test` error is unrelated.  Did you set a FIDO2 PIN?  In that case, you have to call `nitropy nk3 test --pin <yourpin>` to fix the test failure.

Does the update work now?  If not, what are your current udev rules?

**Comment by szszszsz** [comment]
Perhaps you have not reloaded Udev service after copying the rules? Either that, or the previous set was outdated. Anyway, it works now.
@robin-nitrokey We should try to extend the Udev diagnostics in pynitrokey with some semi-automatic procedure for such occasions like this one. 

Leaving ticket open until correcting the initial UI error message.

**Comment by szszszsz** [comment]
Corrected through #195. Closing as done.
